File: ihchat/app/canais/registro.py
"""Fabrica de adaptadores a partir de um canal cadastrado."""
from __future__ import annotations

from ..models import Canal, TipoCanal
from .base import AdaptadorCanal
from .email import AdaptadorEmail
from .telegram import AdaptadorTelegram
from .webchat import AdaptadorWebchat
from .whatsapp import AdaptadorWhatsApp
from .whatsapp_qr import AdaptadorWhatsAppQR

ADAPTADORES: dict[str, type[AdaptadorCanal]] = {
    TipoCanal.WHATSAPP.value: AdaptadorWhatsApp,
    TipoCanal.WHATSAPP_QR.value: AdaptadorWhatsAppQR,
    TipoCanal.TELEGRAM.value: AdaptadorTelegram,
    TipoCanal.EMAIL.value: AdaptadorEmail,
    TipoCanal.WEBCHAT.value: AdaptadorWebchat,
}


class CanalNaoSuportado(Exception):
    pass


def adaptador_para(canal: Canal) -> AdaptadorCanal:
    classe = ADAPTADORES.get(canal.tipo)
    if classe is None:
        raise CanalNaoSuportado(f"tipo de canal desconhecido: {canal.tipo}")
    return classe(canal)

File: omnichannel/php/app/Canais/Registro.php
<?php
declare(strict_types=1);

namespace OmniChannel\Canais;

/** Fábrica de adaptadores a partir de um canal cadastrado (app/canais/registro.py). */
final class Registro
{
    public const ADAPTADORES = [
        Campos::WHATSAPP => AdaptadorWhatsApp::class,
        Campos::TELEGRAM => AdaptadorTelegram::class,
        Campos::EMAIL => AdaptadorEmail::class,
        Campos::WEBCHAT => AdaptadorWebchat::class,
    ];

    /**
     * @param array<string, mixed> $canal linha tipada (Canais::tipar)
     * @throws CanalNaoSuportado
     */
    public static function adaptadorPara(array $canal): Adaptador
    {
        $classe = self::ADAPTADORES[$canal['tipo'] ?? ''] ?? null;
        if ($classe === null) {
            throw new CanalNaoSuportado('tipo de canal desconhecido: ' . ($canal['tipo'] ?? ''));
        }
        return new $classe($canal);
    }
}
